src/ast/passes/semantic_analyser.cpp

```diff
@@ -2841,7 +2841,7 @@ void SemanticAnalyser::visit(MapAccess &acc)
   } else {
     // If there is no record of any assignment after the first pass
     // then it's safe to say this map is undefined.
-    if (!is_first_pass()) {
+    if (!is_first_pass() && !acc.map->read_only) {
```

Comment:
This feels a bit strange, and suggests that you want to actually insert this in the known set of maps for semantic analysis or do something else.